Merge rp-process branch into development

KiranEiden opened the pull request:
**This pull request introduces the following changes:**

- Added `gridplot` function for plotting nuclides on a grid and coloring the cells by X, Y, Xdot, Ydot or activity (and updated pynucastro examples accordingly). Also added evaluation functions for Ydot and activity without electron screening.

- Created a notebook and a script for generating rp-process networks. The script allows an arbitrary endpoint to be specified.

- Added a `filter_function` option to `RateFilter` and to the plotting functions that allows for filtering rates or nuclei using a user-defined Boolean map.

- Changed the the hash function of the `Nucleus` class to use `hash((self.Z, self.A))` instead of `hash(repr(self))`. This lets `Nucleus('p')` be interchangeable with `Nucleus('h1')` as a dictionary key, and ensures that `if Nucleus('A') == Nucleus('B'): hash(Nucleus('A')) == hash(Nucleus('B'))`. I also updated the equals method to allow comparison to tuples, which lets us use `(Z, A)` tuples as dict keys: e.g. `dict[2, 4]` instead of `dict[Nucleus('he4')]`.

- Currently a `write_network` call will crash if the `RateCollection` contains multiple rates with the same name. I added an option for specifying a preference for one rate label over another when constructing the `RateCollection`, which lets it automatically trim out the extra rates. I use it in [rp_process.py](https://github.com/pynucastro/pynucastro/blob/rp-process/examples/rp-process/rp_process.py) to give wc17 rates precedence over ths8 ones.

- Added `numba` to the `install_requires` list. #102 fixed the `DeprecationWarning` raised by the `jitclass` import -- this wraps the import in a try-catch to still support older versions of `numba`.

**Some notes:**

- I added and removed a 300k line reaclib library file, so this pull request needs to be squashed. I don't know if that will be done automatically.

- Because I removed the file, the Travis CI build will fail when it tries to run [network.ipynb](https://github.com/pynucastro/pynucastro/blob/rp-process/examples/rp-process/network.ipynb), which requires the full library. I can remove the notebook, as its functionality is duplicated in the script, but I thought that it would be nice to have a notebook form of the example. Is it possible to have the Travis build ignore that particular file? **I added a nbval-skip tag to each cell -- I'm not sure if that will do the trick or not.**

- The `Nucleus` class uses the raw input string passed to the constructor for its `repr`. It may be better to use the short element name to give the class a more standardized representation, but I would need to rerun all of the notebooks if I switched to that because the output would change.

dwillcox reviewed (changes requested):
Overall this looks great! There's so much incredibly elegant code here and much-needed features for pynucastro, and I've noted just a few places throughout the code for some minor improvements.

One thing to fix is a rate deletion bug in the `_make_distinguishable` function of `rate_collection.py`, and the review comment has a proposed fix. Please test & add to the PR :-)

KiranEiden commented:
Hi Don,

Thank you for taking the time to do a detailed review of this! I've updated the pull request to incorporate the changes that you requested, along with one or two other things:

- Your fix for the bug in the `_make_distinguishable` function seems to work well -- I've introduced it to the function implementation. I've also updated the comment in the `evaluate_activity` function.
- The `rp_process.py` script now uses `argparse`, and can take in both the endpoint of the network and the library file name. I added some print statements so it gives some indication of progress as it runs.
- I reran `pynucastro-examples.ipynb` and had it save the widget state (I think I forgot to do this last time), so the widget should show up properly now.
- In rerunning the notebook, I found that the previous implementation of the `plot` function in `RateCollection` doesn't work with networkx 2.5.1 -- some keyword arguments like `zorder` are ostensibly no longer valid. I solved this by wrapping the calls in a try-except clause, and supplying different arguments if the `except` is triggered.
- I left the electron capture rates in, as I'm not sure how to filter them out and it doesn't seem like it will hurt to have them in the network. If you happen to know a way of distinguishing beta+ decays from electron captures in ReacLib and think we should remove the latter, we definitely can though.

dwillcox commented on examples/rp-process/rp_process.py:
```diff
@@ -1,21 +1,40 @@
+library_help = """The library file to draw the rates from. This is supplied directly to the Library
+        constructor."""
+
+parser = argparse.ArgumentParser(description=description)
```
This looks great, thanks!

dwillcox commented:
Thanks very much @KiranEiden ! These changes are perfect :)

I think it's fine to leave the electron capture rates in, just as long as we're aware of it.

If in the future we really want to get electron captures right, though, we'll need to use the weak rates tabulated in `(T, rho*Ye)` instead of the Reaclib-provided electron capture rates.

And as a bonus, this closes #111
